### ISSUE TITLE
Compatibility problem on MultiSelect with KendoUI version 2015.2.624

### ISSUE THREAD

**Issue by bzamfir** [kendo-labs/knockout-kendo#208]
After updating to KendoUI 2015.2.624, there is a compatibility problem to MultiSelect control

I created two jsfiddles to exemplify the problem:

With KendoUI 2015.1.429 the MultiSelect works ok
http://jsfiddle.net/bzamfir/3ng6yxxk/

Starting with KendoUI 2015.2.624, MultiSeletc no longer works
http://jsfiddle.net/bzamfir/b754rgor/

Please check and fix if possible.
Thanks you

**Comment by bmsmg** [comment]
I reported a similar issue in #202.  Kendo changed the underlying object
used for working with lists.  You will need to stick with the required
version for ko-kendo 0.9.3, which I believe is 2014.3.

On Mon, Aug 3, 2015 at 3:13 PM, bzamfir <notifications@github.com> wrote:

> After updating to KendoUI 2015.2.624, there is a compatibility problem to
> MultiSelect control
>
> I created two jsfiddles to exemplify the problem:
>
> With KendoUI 2015.1.429 the MultiSelect works ok
> http://jsfiddle.net/bzamfir/3ng6yxxk/
>
> Starting with KendoUI 2015.2.624, MultiSeletc no longer works
> http://jsfiddle.net/bzamfir/b754rgor/
>
> Please check and fix if possible.
> Thanks you
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kendo-labs/knockout-kendo/issues/208>.
>



-- 
*Matthew Grant* | *Sr. Software Engineer* | Black Mountain Systems |
858.461.6706 | mgrant@blackmountainsystems.com

**Comment by bzamfir** [comment]
I reported the issue to kendoui support
Hopefully they will fix the issue

2015-08-04 1:45 GMT+03:00 bmsmg <notifications@github.com>:

> I reported a similar issue in #202. Kendo changed the underlying object
> used for working with lists. You will need to stick with the required
> version for ko-kendo 0.9.3, which I believe is 2014.3.
>

**Comment by bzamfir** [comment]
The answer from Telerik support was:
> According to my observation the error occurs in Kendo-Knockout source files. I would like to remind > you that Kendo-Knockout is an open source project which is not developed nor maintained by the > Kendo UI team.  As such, this project is not officially supported by the Kendo UI team.
> Any support requests (via tickets or Forum posts) should go to either StackOverflow, or to the Issues > list for that project.

I replied them that, since KO respectively ko-knockout remain unchanged, the "breaking change" (read: bug) was introduced in KendoUI. But I'm not sure they will take any action.

I also noticed to issues reported here on the same matter: #201 and #202.

Any idea, if the compatibility can be fixed in ko-kendo lib, or is related better to KO internally?

**Comment by bmsmg** [comment]
Knockout-Kendo is a 3rd party utility that makes working with both packages
much easier.  Ryan Niemeyer, the author of Knockout, is not affiliated with
Kendo.  He just saw a problem and solved it.  Your options are 1.) to wait
until someone decides to figure out what Kendo has changed in their newest
version and modify Knockout-Kendo accordingly, 2.) make a pull request and
do this yourself, 3.) switch to Angular as Kendo officially supports it
(though, Kendo used to support Knockout as well - so beware dependencies).

On Tue, Aug 4, 2015 at 4:00 AM, bzamfir <notifications@github.com> wrote:

> The answer from Telerik support was:
>
> According to my observation the error occurs in Kendo-Knockout source
> files. I would like to remind > you that Kendo-Knockout is an open source
> project which is not developed nor maintained by the > Kendo UI team. As
> such, this project is not officially supported by the Kendo UI team.
> Any support requests (via tickets or Forum posts) should go to either
> StackOverflow, or to the Issues > list for that project.
>
> I replied them that, since KO respectively ko-knockout remain unchanged,
> the "breaking change" (read: bug) was introduced in KendoUI. But I'm not
> sure they will take any action.
>
> I also noticed to issues reported here on the same matter: #201
> <https://github.com/kendo-labs/knockout-kendo/issues/201> and #202
> <https://github.com/kendo-labs/knockout-kendo/issues/202>.
>
> Any idea, if the compatibility can be fixed in ko-kendo lib, or is related
> better to KO internally?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kendo-labs/knockout-kendo/issues/208#issuecomment-127565874>
> .
>



-- 
*Matthew Grant* | *Sr. Software Engineer* | Black Mountain Systems |
858.461.6706 | mgrant@blackmountainsystems.com

**Comment by bzamfir** [comment]
Switching to angular is not an option because we already have a lot of modules developed with ko-kendo.
And I don't know enough about ko internals to be able to fix ko-kendo lib quickly.

But finally I might have to do it we I'll be forced to update to a newer kendoui version and if Telerik don't fix it.

**Comment by rniemeyer** [comment]
@bmsmg @bzamfir - sorry, I haven't had a chance to look at this issue to see if there is some type of easy fix

**Comment by bzamfir** [comment]
@rniemeyer, thanks for taking time to check and reply on this.
The issue is not top emergency (for me at least), as I can stay with 2015.1.429 for now (actually, with v2015.1.528 - an internal build based on official 2015.1.429, which still works ok).
The error was stared sometime after that version.  

I insisted to Telerik team to check the issue themselves.
But if you can get some time to check and update ko-kendo lib for compatibility with current kendoui, that would be great, and I will be grateful.

Cordially,
B. Zamfir

**Comment by bzamfir** [comment]
Actually I figured out what the problem is (at least in my case).

Using KendoUI 2015.2.624 (first one with breaking change)
ko-kendo lib with Multi-select - http://jsfiddle.net/b754rgor/2/ - works ok
ko-kendo lib with Multi-select - http://jsfiddle.net/bzamfir/b754rgor/4/ - doesn't work

The difference is, the one that works uses <select> as HTML bound element, while the one that doesn't work uses <input> as bound element

Even if <input> used to work ok when bound to MultiSelect pre-v2015.2.624, it no longer works starting with v2015.2.624

Just hoping that this will help someone facing the same issue.

**Comment by alexeysuvorov** [comment]
Looks like the same error for TabStrip.

**Comment by IanYates** [comment]
I just upgraded to 2015.2 a couple of days ago and can confirm I ran into this issue - but fortunately knew the fix thanks to this thread :)

I had `<input data-bind="kendoMultiSelect:{.....}" />`
Changing that to `<select>` tags fixed the problem.  I was previously on 2015.1.  I suspect where our site was using the `<input>` tags previously correlates with reports of multiselects not working IE 11 previously (anecdotally they now all definitely work in IE 11 whereas before I'd seen them not work sometimes)

@alexeysuvorov  - I haven't used tabstrip in our site, but if I ever do it'd be good to know the fix if you ever find one.  What HTML element are you using at the moment?  Could you try changing it to something else (div?) and see if that helps?

**Comment by alexeysuvorov** [comment]
I made it works for me by replacing line 145 in knockout-kendo-core.js (just remove colon from regexp).
from 
    var parent = $element.closest("[data-bind*='" + widgetConfig.parent + ":']");
to 
    var parent = $element.closest("[data-bind*='" + widgetConfig.parent + "']");
My binding looks like 
    kendoTabStrip : { animation : false } //space after colon

**Comment by rniemeyer** [comment]
@alexeysuvorov - Ah thanks. I will look to improve that code.

**Comment by saurabh0jha** [comment]
Are there any knockout-kendo upgrades in pipeline  to support newer versions of Kendo UI.?The information will help in choosing the newer kendo version to safely upgrade to.

I am facing the same issue. I upgraded kendoUI to the latest version as of now (**kendo 2015.2.902**). While most of the widgets i had used were working fine only **kendoMultiSelect** breaks. So i am restricted to upgrade kendo UI to the knockout-kendo supported version.

**Comment by srowan** [comment]
Has anyone found a workaround or patch for this?  IanYates suggestion of using a select tag did nothing for us.

**Comment by bmsmg** [comment]
I think I have a *fix* for the MultiSelect issue.  I found the problem at line 226 of knockout-kendo.js v0.9.3.  There is still some investigation necessary, so this may be simply treating a symptom, but hopefully it helps continue the conversation.

Problem manifests here:

```javascript
//try to avoid unnecessary updates when the new value matches the current value
if (custom || existing !== value) {
	action.apply(widget, params);
}
```
The issue is the inequality comparison between 'existing' and 'value'.  Since they can be arrays in the MultiSelect case, a simple ===/!== will not work.  So, this bit of code will ALWAYS be called (note the comment above about avoiding unnecessary updates).  The value has already been set a few lines above.  On the Kendo side, this ends up messing with the list control's data view, which results in _valueIndices returning just the most recent value.

I made the following changes on the knockout-kendo side to address the equality issue:

[1].  Added equality comparison functions that can handle comparison of arrays.  I borrowed this from Underscore's _.isEqual function, but there may be a simpler choice.  I kept these in the scope of the BindingFactory for now.

```javascript
//Underscore.js 1.8.3
//http://underscorejs.org
//(c) 2009-2015 Jeremy Ashkenas, DocumentCloud and Investigative Reporters & Editors
//Underscore may be freely distributed under the MIT license.
// Internal recursive comparison function for `isEqual`.
var eq = function ( a, b, aStack, bStack ) {
	// Identical objects are equal. `0 === -0`, but they aren't identical.
	// See the [Harmony `egal` proposal](http://wiki.ecmascript.org/doku.php?id=harmony:egal).
	if ( a === b ) return a !== 0 || 1 / a === 1 / b;
	// A strict comparison is necessary because `null == undefined`.
	if ( a == null || b == null ) return a === b;
	// `NaN`s are equivalent, but non-reflexive.
	if ( a !== a ) return b !== b;
	// Exhaust primitive checks
	var type = typeof a;
	if ( type !== 'function' && type !== 'object' && typeof b != 'object' ) return false;
	return deepEq( a, b, aStack, bStack );
};

// Internal recursive comparison function for `isEqual`.
var deepEq = function ( a, b, aStack, bStack ) {
	// Unwrap any wrapped objects.
	if ( a instanceof _ ) a = a._wrapped;
	if ( b instanceof _ ) b = b._wrapped;
	// Compare `[[Class]]` names.
	var className = toString.call( a );
	if ( className !== toString.call( b ) ) return false;
	switch ( className ) {
		// Strings, numbers, regular expressions, dates, and booleans are compared by value.
		case '[object RegExp]':
			// RegExps are coerced to strings for comparison (Note: '' + /a/i === '/a/i')
		case '[object String]':
			// Primitives and their corresponding object wrappers are equivalent; thus, `"5"` is
			// equivalent to `new String("5")`.
			return '' + a === '' + b;
		case '[object Number]':
			// `NaN`s are equivalent, but non-reflexive.
			// Object(NaN) is equivalent to NaN
			if ( +a !== +a ) return +b !== +b;
			// An `egal` comparison is performed for other numeric values.
			return +a === 0 ? 1 / +a === 1 / b : +a === +b;
		case '[object Date]':
		case '[object Boolean]':
			// Coerce dates and booleans to numeric primitive values. Dates are compared by their
			// millisecond representations. Note that invalid dates with millisecond representations
			// of `NaN` are not equivalent.
			return +a === +b;
	}

	var areArrays = className === '[object Array]';
	if ( !areArrays ) {
		if ( typeof a != 'object' || typeof b != 'object' ) return false;

		// Objects with different constructors are not equivalent, but `Object`s or `Array`s
		// from different frames are.
		var aCtor = a.constructor, bCtor = b.constructor;
		if ( aCtor !== bCtor && !( _.isFunction( aCtor ) && aCtor instanceof aCtor &&
								 _.isFunction( bCtor ) && bCtor instanceof bCtor )
							&& ( 'constructor' in a && 'constructor' in b ) ) {
			return false;
		}
	}
	// Assume equality for cyclic structures. The algorithm for detecting cyclic
	// structures is adapted from ES 5.1 section 15.12.3, abstract operation `JO`.

	// Initializing stack of traversed objects.
	// It's done here since we only need them for objects and arrays comparison.
	aStack = aStack || [];
	bStack = bStack || [];
	var length = aStack.length;
	while ( length-- ) {
		// Linear search. Performance is inversely proportional to the number of
		// unique nested structures.
		if ( aStack[length] === a ) return bStack[length] === b;
	}

	// Add the first object to the stack of traversed objects.
	aStack.push( a );
	bStack.push( b );

	// Recursively compare objects and arrays.
	if ( areArrays ) {
		// Compare array lengths to determine if a deep comparison is necessary.
		length = a.length;
		if ( length !== b.length ) return false;
		// Deep compare the contents, ignoring non-numeric properties.
		while ( length-- ) {
			if ( !eq( a[length], b[length], aStack, bStack ) ) return false;
		}
	} else {
		// Deep compare objects.
		var keys = _.keys( a ), key;
		length = keys.length;
		// Ensure that both objects contain the same number of properties before comparing deep equality.
		if ( _.keys( b ).length !== length ) return false;
		while ( length-- ) {
			// Deep compare each member
			key = keys[length];
			if ( !( _.has( b, key ) && eq( a[key], b[key], aStack, bStack ) ) ) return false;
		}
	}
	// Remove the first object from the stack of traversed objects.
	aStack.pop();
	bStack.pop();
	return true;
};
```

[2]. I replaced the inequality condition at line 226 with this:

```javascript
//try to avoid unnecessary updates when the new value matches the current value
var equalTo = eq(existing, value);
if ( custom || !equalTo ) {
	action.apply( widget, params );
}
```

I still need to compare the states of the data between this release - 2015.3 - and the 2014.3 kendo release where this used to work to see if the issue actually lies elsewhere.  Also, kendo switched to an underlying StaticList control after 2014.3 which may require a change in the flow of the watchOneValue function (?).

**Comment by IanYates** [comment]
@bmsmg - great detective work.

Your explanation makes sense, and also may explain why I haven't really seen the problem in our codebase for a while.  Most of the data sources I use for a multi-select component tend to be derived in a computed observable, so it's a new array instance each time.

I'd be keen to hear what others think...

**Comment by rniemeyer** [comment]
I added a change in 0.9.6 to properly clear the filter when the value is updated for kendoMultiSelect. In my testing, this seems to address any issues with the box being out of sync with the selected values from either selecting additional values or by programmatically setting the selected values to something else. I did not see a quick way to make it work with an `input` element. A `select` element should be used with `kendoMultiSelect` binding. 

Please let me know if anyone sees additional problems/issues.

**Comment by bzamfir** [comment]
Hi,

I'm using ko-kendo through nuget.
Can you update and publish the new version as nuget package?

Thank you
B. Zamfir


2015-11-14 18:25 GMT+02:00 Ryan Niemeyer <notifications@github.com>:

**Comment by rniemeyer** [comment]
@bzamfir published 0.9.6 to Nuget. Thanks.

**Comment by bzamfir** [comment]
Thank you

2015-11-16 16:17 GMT+02:00 Ryan Niemeyer <notifications@github.com>:
